File: backend/app/schemas/vote.py
from pydantic import BaseModel, Field

from app.models.enums import SectionType, VoteType


class VoteRequest(BaseModel):
    """Body for POST /vote – cast or update a vote. One vote per (user, section_type, item_id)."""

    section_type: SectionType = Field(
        ...,
        description="Dashboard section: news, price, ai, or meme",
    )
    item_id: str = Field(
        ...,
        min_length=1,
        max_length=255,
        description="External content identifier (e.g. news id, coin symbol, meme id)",
    )
    vote_type: VoteType = Field(
        ...,
        description="up or down",
    )


class VoteResponse(BaseModel):
    """Response after casting or updating a vote."""

    status: str = "ok"
    action: str = Field(
        ...,
        description="created (new vote) or updated (existing vote changed)",
    )


class VoteCancelRequest(BaseModel):
    """Body for DELETE /vote – remove a vote (cancel)."""

    section_type: SectionType = Field(
        ...,
        description="Dashboard section: news, price, ai, or meme",
    )
    item_id: str = Field(
        ...,
        min_length=1,
        max_length=255,
        description="External content identifier for the vote to cancel",
    )


class VoteCancelResponse(BaseModel):
    """Response after cancelling a vote."""

    status: str = "ok"
    action: str = "cancelled"

File: frontend/src/api/vote.ts
import { apiClient } from './client'
import type { VoteCancelRequest, VoteRequest } from '../types/vote'

export interface VoteResponse {
  status: string
  action: 'created' | 'updated'
}

export interface VoteCancelResponse {
  status: string
  action: string
}

export async function postVote(body: VoteRequest): Promise<VoteResponse> {
  const { data } = await apiClient.post<VoteResponse>('/vote', body)
  return data
}

export async function deleteVote(body: VoteCancelRequest): Promise<VoteCancelResponse> {
  const { data } = await apiClient.delete<VoteCancelResponse>('/vote', { data: body })
  return data
}
